openedx/core/djangoapps/edx_global_analytics/tests/test_cache_instance_data.py

```diff
@@ -0,0 +1,100 @@
+"""
+Tests for edX global analytics application cache functionality.
+"""
+
+from datetime import date, datetime
+
+from mock import patch
+
+from django.test import TestCase
+from django.utils import timezone
+from django.db.models import Q
+
+from student.models import UserProfile
+from student.tests.factories import UserFactory
+
+from openedx.core.djangoapps.edx_global_analytics.utils import (
+    cache_instance_data,
+    cache_timeout_month,
+    cache_timeout_week,
+)
+
+
+class TestCacheInstanceData(TestCase):
+    """
+    Test cover cache-functionality for queries results.
+    """
+
+    @staticmethod
+    def create_default_data():
+        """
+        Default integration database data for cache instance information tests.
+        """
+        users_last_login = [
+            timezone.make_aware(datetime(2017, 5, 8, 0, 0, 0), timezone.get_default_timezone()),
+            timezone.make_aware(datetime(2017, 5, 14, 23, 59, 59), timezone.get_default_timezone()),
+            timezone.make_aware(datetime(2017, 5, 15, 0, 0, 0), timezone.get_default_timezone()),
+            timezone.make_aware(datetime(2017, 5, 15, 0, 0, 1), timezone.get_default_timezone())
+        ]
+
+        for user_last_login in users_last_login:
+            UserFactory(last_login=user_last_login)
+
+    def test_cache_new_instance_data(self):
+        """
+        Verify that cache_instance_data returns data as expected after caching it.
+        """
+        self.create_default_data()
+
+        period_start, period_end = date(2017, 5, 8), date(2017, 5, 15)
+
+        active_students_amount_week = UserProfile.objects.exclude(
+            Q(user__last_login=None) | Q(user__is_active=False)
+        ).filter(user__last_login__gte=period_start, user__last_login__lt=period_end).count()
+
+        cache_timeout = None
+
+        result = cache_instance_data('active_students_amount_week', active_students_amount_week, cache_timeout)
+
+        self.assertEqual(result, 2)
```

Comment:
Because it include `0:0:0` of start of day and exclude `0:0:0` of end of day.

`date(2017, 5, 8)` -> `(2017, 5, 8, 0, 0, 0)` -> `user__last_login__gte=period_start` -> include.
`date(2017, 5, 15)` -> `(2017, 5, 15, 0, 0, 0)` -> `user__last_login__lt=period_end` -> exclude.